### DISCUSSION TITLE
Can we modify to be imported module name before auto completion?

### DISCUSSION THREAD

**Discussion by wztdream** [company-mode/company-mode#1220] [Ideas]
company-mode is great!
Suppose we need import `numpy` during coding, current work flow is:
1. user type `numpy`
2. trigger auto completion `import numpy`
3. user choose `import numpy`
4. auto insert `import numpy` to the file, and insert `numpy` at point`

But usually we need `import numpy as np` not `import numpy` so the handy work flow is:

1. user type `numpy`
2. trigger auto completion `import numpy`
3. user chose `import numpy` by `tab`
4. user edit `import numpy` to `import numpy as np`, then `enter`
5. auto insert `import numpy as np` to the file, and insert `np` at the point 

this will avoid user to manually change `import numpy` to `import numpy as np` which need to scrool up and down to find the correct location.

**Comment by dgutov** [comment]
This is a feature that needs to be implemented or configured in the backend. See `M-x company-diag` to find out the one you're using.

I think (not sure) LSP servers have something like that already.

**Reply by dgutov** [reply]
Or if you are using the built-in Python completion (which is based on the inferior shell), this can be a feature request for Emacs core.

**Reply by wztdream** [reply]
Here is the output from `M-x company-diag`, lsp can do auto import as I described in the first work flow, but not the desired workflow(the second work flow). I will ask LSP mode side then,  I suppose this is related with the server of 'lsp-mode' which is 'pyright' in my case.
Any way, for company, is there any hook I can use to do some hacking in case the server not support such action?
```
company-backends: (company-capf)

Used backend: company-capf
Value of c-a-p-f: (lsp-completion-at-point)
Major mode: python-mode
```

**Reply by dgutov** [reply]
> I suppose this is related with the server of 'lsp-mode' which is 'pyright' in my case.

Yup.

> Any way, for company, is there any hook I can use to do some hacking in case the server not support such action?

Not in this case, probably. Adding a hook or advice is easy. But how will your code know when it needs to do this, and that `numpy` is a package name to be imported and not, say, a simple identifier?

**Reply by wztdream** [reply]
This features is indeed provided by `pylance` but unfortunately pylance is not available in emacs and the owner of `pyright`(which is available to emacs) is not planning to support this function.   [#1837](https://github.com/microsoft/pyright/issues/1837)

**Reply by dgutov** [reply]
Is that the preferred LS for Python? I can see lsp-mode's docs also mention

https://emacs-lsp.github.io/lsp-python-ms/
and
https://emacs-lsp.github.io/lsp-mode/page/lsp-jedi/

**Reply by wztdream** [reply]
> Is that the preferred LS for Python? I can see lsp-mode's docs also mention
> 
> https://emacs-lsp.github.io/lsp-python-ms/
Microsoft python language server will be replaced by new pylance for the long term, so there is limited maintain for the old one.
> https://emacs-lsp.github.io/lsp-mode/page/lsp-jedi/
I have not use jedi, maybe I will have a try.

**Comment by wztdream** [comment]
`Not in this case, probably. Adding a hook or advice is easy. But how will your code know when it needs to do this, and that numpy is a package name to be imported and not, say, a simple identifier?`

As current `pyright` already support the workfollow(1st), so we can get the imperfect auto import `import numpy` in this example, so my naive idea is to `manually` modify the result of `pyright` from `import numpy` to `import numpy as np` as described in the second workflow. But to do so, I need figure out where to hack, maybe I need some help from lsp-mode?

**Reply by dgutov** [reply]
But how will you determine whether to modify the result?

There are different places your can try. `company-transformers`, for example. But you probably need to modify the buffer before the completion request is sent. Then it's better performed as a `:before` advice on `company-auto-begin`. Or you can even use `pre-command-hook`.

**Reply by wztdream** [reply]
`But how will you determine whether to modify the result?`

This is determined by the user, say, I want to rename `numpy` as `np` or maybe `nu`, this is user dependent, the **auto** part is: 
1. generate candidate **standard** module name, such as `numpy`, 
2. find the correct location to insert the code.

 The manual part is rename the module name.

**Reply by dgutov** [reply]
Sounds like you want to write a new command called, say, `my/python-rename-identifier`?

**Reply by wztdream** [reply]
> Sounds like you want to write a new command called, say, `my/python-rename-identifier`?

yes, maybe something like that, but I will first try some new server like jedi to see if they support this function already
